internal/indexer/oci.go

```diff
@@ -0,0 +1,201 @@
+package indexer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/bitnami-labs/charts-syncer/internal/indexer/api"
+	"github.com/bitnami-labs/pbjson"
+	containerderrs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes"
+	"github.com/pkg/errors"
+	"oras.land/oras-go/pkg/content"
+	"oras.land/oras-go/pkg/oras"
+)
+
+// ociIndexerOpts are the options to configure the ociIndexer
+type ociIndexerOpts struct {
+	reference string
+	url      string
+	username string
+	password string
+	insecure bool
+}
+
+// OciIndexerOpt allows setting configuration options
+type OciIndexerOpt func(opts *ociIndexerOpts)
+
+// WithIndexRef configures the charts index OCI reference instead of letting the library
+// using the default host/index:latest one.
+//
+// 	opt := WithIndexRef("my.oci.domain/index:prod")
+//
+func WithIndexRef(r string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.reference = r
+	}
+}
+
+// WithBasicAuth configures basic authentication for the OCI host
+//
+// 	opt := WithBasicAuth("user", "pass")
+//
+func WithBasicAuth(user, pass string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.username = user
+		opts.password = pass
+	}
+}
+
+// WithInsecure configures insecure connection
+//
+// 	opt := WithInsecure()
+//
+func WithInsecure() OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.insecure = true
+	}
+}
+
+// WithHost configures the OCI host
+//
+// 	opt := WithHost("my.oci.domain")
+//
+func WithHost(h string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.url = h
+	}
+}
+
+// ociIndexer is an OCI-based Indexer
+type ociIndexer struct {
+	reference string
+	resolver  remotes.Resolver
+}
+
+// DefaultIndexName is the name for the OCI artifact with the index
+const DefaultIndexName = "index"
+
+// DefaultIndexTag is the tag for the OCI artifact with the index
+const DefaultIndexTag = "latest"
+
+// NewOciIndexer returns a new OCI-based indexer
+func NewOciIndexer(opts ...OciIndexerOpt) (Indexer, error) {
+	opt := &ociIndexerOpts{}
+	for _, o := range opts {
+		o(opt)
+	}
+
+	u, err := url.Parse(opt.url)
+	if err != nil {
+		return nil, errors.Wrapf(ErrInvalidArgument, "invalid OCI host URL: %+v", err)
+	}
+	resolver := newDockerResolver(u, opt.username, opt.password, opt.insecure)
+
+	ind := &ociIndexer{
+		reference: opt.reference,
+		resolver:  resolver,
+	}
+	if ind.reference == "" {
+		host := strings.Trim(strings.Join([]string{u.Host, u.Path}, "/"), "/")
```

Comment:
This is not a proper host right? it also includes the project name and subpath? 
In that case I would avoid naming the variable as `host`, maybe `partialReference` or `partialPath`?